### PR TITLE
Merge TCP and UDP ports into a single service

### PR DESCRIPTION
**Pull request by jnoordsij**
### What does this PR do?

Merge TCP and UDP ports into a single service by default. Since Kubernetes 1.20, mixed protocol loadbalancers are supported. 

This is partly based on the work in https://github.com/traefik/traefik-helm-chart/pull/420.

### Motivation

The main goal is to allow an easy way to use HTTP/3 with a single load balancer using this chart.

### More

- [x] Yes, I updated the [chart version](https://github.com/traefik/traefik-helm-chart/blob/9b32ed1b414cc0be1ad46bcb335fcfc93ded06f3/traefik/Chart.yaml#L5)

### Additional Notes

This is a work in progress; I have yet to
- [x] add additional tests
- [x] try and test this in practice somehow

I'm mostly looking for feedback to my approach and to users having attempted to try similar things in the past and/or are currently using UDP ports or HTTP/3.

### CODE REVIEW

**Comment by tobru** [comment]
My use-case would be to route DNS traffic through Traefik which needs TCP and UDP on the same IP address, so I'd love to see this merged into one service :+1:

**Comment by jnoordsij** [comment]
Okay, I've managed to apply this and use HTTP/3 succesfully on my Traefik running on a DigitalOcean managed Kubernetes cluster, by simply passing the `experimental.http3 = true` and `ports.websecure.http3.exposed = true` flags. Tests are all fine to me. Ready for review!

**Review by mloiseleur — CHANGES_REQUESTED** [review]

File: traefik/templates/service.yaml [hunk 1]
```diff
@@ -1,25 +1,33 @@
...
+    {{/* Due to https://github.com/kubernetes/kubernetes/issues/58477 it is not possible
+      to define the HTTP/3 UDP port as a (named) containerPort. TCP and UDP containerPort
+      will have the same value for 'port' which will break the chart upgrade. Therefore
+      we manually refer to port number for HTTP/3. */}}
+    {{- $http3Name := (toString $config.port) }}
+    {{- $_ := set $udpPorts $http3Name $config.http3 -}}
+  {{- end -}}
+  {{- if eq (toString $config.protocol) "UDP" -}}
+    {{ $_ := set $udpPorts $name $config -}}
+  {{- end -}}
```
[the remark this answers is not in the document]
Comment (jnoordsij) [reply traefik/templates/service.yaml]: I think this mostly comes down to how you want to structure the port values in `values.yaml`.
The `http3` config is now nested within `websecure`, because they are so tightly related.

This causes it not to be considered when listing container-ports on the pod, which is neccessary, as listing two ports with the same port number is currently not possible (in a reliable way) in Kubernetes due to merging mechanics. Therefore it is not a named pod-port like the others (e.g. `web` or `websecure`), so it requires a reference by port number instead of port name.

For all other ports, including other UDP-ports, there is no problem in referring to them by name. You would still need to be careful not to define multiple named ports that share the same port number, but I think that's practically a different problem. Disabling port-by-name

File: traefik/Changelog.md [hunk 2]
```diff
@@ -1,5 +1,57 @@
...
++      # The port protocol (TCP/UDP)
++      protocol: UDP
+     # Set TLS at the entrypoint
+     # https://doc.traefik.io/traefik/routing/entrypoints/#tls
+     tls:
+@@ -500,6 +506,7 @@ tlsStore: {}
+ # from.
+ service:
+   enabled: true
++  single: true
```
[the remark this answers is not in the document]
Comment (jnoordsij) [reply traefik/Changelog.md:51]: Yeah, I had some doubts about it myself as well. Basically my argumentation is: this default seems to be both the most-used and most-reasonable value. 

The old value creates 2 separate services, which will result in 2 different LBs. Although some people might actually want to do that, I think most people starting out with this chart will be utterly confused (like me) that enabling HTTP/3 causes an additional LB to spin up. Moreover, they'll probably revert in this case, as routing HTTP/3 through a different LB is even more advanced and more costly as well.

With the new value, people trying to enable HTTP/3 on a non-supporting cluster will simply run into a Kubernetes warning/error telling them there is no support for their service, which feels more 'logical' to me.

You could consider adding some sort of service-detection for multi-protocol support, and warning users in advance.

[the remark this answers is not in the document]
Comment (jnoordsij) [reply traefik/Changelog.md:51]: Also note that the multi-protocol feature has been available since Kubernetes 1.20 (see https://github.com/kubernetes/kubernetes/pull/94028) behind a feature-flag; it graduated to beta in 1.24. So there may be more providers who support this.

File: traefik/templates/service.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tobru) [reply traefik/templates/service.yaml]: My use-case does not involve any HTTP/3, but DNS. Would this be supported with the proposed change?

[the remark this answers is not in the document]
Comment (jnoordsij) [reply traefik/templates/service.yaml]: > My use-case does not involve any HTTP/3, but DNS. Would this be supported with the proposed change?

This change does not actually change or add any functionality; all it does is merge all TCP/UDP ports into a single service by default, where previously it would create 2 seperate services. 

The old split-services behavior can be kept with the `service.single: false` value; see also the upgrade notes I added to the README in this PR.

[the remark this answers is not in the document]
Comment (tobru) [reply traefik/templates/service.yaml]: > all it does is merge all TCP/UDP ports into a single service by default, where previously it would create 2 seperate services.

That's perfect for DNS, as it listens on TCP _and_ UDP 53 on the same IP address aka LoadBalancer.

[the remark this answers is not in the document]
Comment (smnbbrv) [reply traefik/templates/service.yaml]: It would be perfect but in K8S it is not possible to listen TCP and UDP on the same port AFAIK. So your use case would be impossible to implement in k8s

[the remark this answers is not in the document]
Comment (jnoordsij) [reply traefik/templates/service.yaml]: Yeah, for http3 this PR explicitly does not use 443 and instead uses an alternative port (8443), as using the same port number for both UDP and TCP [does not fully work in Kubernetes](https://github.com/kubernetes/kubernetes/issues/39188). Creating from scratch will *probably* work, but any edited config might not merge correctly due to this.

File: traefik/Changelog.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mloiseleur) [reply traefik/Changelog.md:51]: You're right, it's possible to enable it since v1.20 and it's not enabled by default before v1.24.
It's still an advanced feature that is available but needs to be enabled on most cloud & kubernetes providers.
Http3 is also an experimental feature. 

Following your arguments, I was thinking about doing both, I mean
1. Disabling single service by default
2. Enabling single service when user enable `http3`

But it means that it won't be possible to use `http3` with two services.
And that's how one can do it currently with this Chart. 
And it works on most cloud & kubernetes providers.

=> So, for this year, I guess we should disable it by default and improve documentation on http3 about this.

[the remark this answers is not in the document]
Comment (yoeluk) [reply traefik/Changelog.md:51]: I was going to open a ticket about something related but noticed this WIP. I like to add that in some providers you need a TCP port open in the NLB for the health check. I came across this because there is no way of deploying the UDP NLB with this chart for 'digitalocean' where you will get this error message "Error syncing load balancer: failed to ensure load balancer: failed to build load-balancer request: specified health check port 80 does not exist on service traefik/traefik-udp". So, it seems odd that to get my single service, and single NLB, I need to enable `http3`. Seems unrelated otherwise.

[the remark this answers is not in the document]
Comment (yoeluk) [reply traefik/Changelog.md:51]: Just for completion. If you attempt to use the same port you opened with UDP protocol you get the following error "Error syncing load balancer: failed to ensure load balancer: failed to build load-balancer request: health check port: 4648, cannot be of protocol type UDP". So, this ticket will solve more than optimising the services into one. For some providers will make it possible to use UDP inbound.

[the remark this answers is not in the document]
Comment (jnoordsij) [reply traefik/Changelog.md:51]: > You're right, it's possible to enable it since v1.20 and it's not enabled by default before v1.24. It's still an advanced feature that is available but needs to be enabled on most cloud & kubernetes providers. Http3 is also an experimental feature.
> 
> Following your arguments, I was thinking about doing both, I mean
> 
> 1. Disabling single service by default
> 2. Enabling single service when user enable `http3`
> 
> But it means that it won't be possible to use `http3` with two services. And that's how one can do it currently with this Chart. And it works on most cloud & kubernetes providers.
> 
> => So, for this year, I guess we should disable it by default and improve documentation on http3 about this.

I do agree support for mixed protocols might be limited (you could actually do a quick check to see how well-supported it is among major providers; I'm using Digital Ocean and they only added UDP support recently), so that might be a reason to keep the old default.

Then again, I think the current default of automatically creating two LB's is even more advanced. My assumption is few people are using this in practice, as it will mean 2 IP's and hence a lot of hassle to use. Actually I'm still curious if there are any user stories of people using the current chart for http3, hence my initial call for feedback.

Moreover the result of suddenly having two load balancers is confusing and potentially dangerous (I ended up accidentally deleting my LB while rollbacking my chart the first time I tried).

So basically I think the old default only already serves quite advanced (edge) cases. 

In the end it's your call what's best for now, happy to change back to the old value if requested.

Side note: I will probably not have any time to adjust things this week, so I'll look further into any requested changes next week.

[the remark this answers is not in the document]
Comment (smnbbrv) [reply traefik/Changelog.md:51]: It must be just dumb easy. One Boolean value more or less - is fine. I can recall traefik in its early days already supporting let's encrypt and http2 and that was why we chose it. I tried to configure traefik with MetalLB and two services last week with current chart and I literally was reading the chart sources to find out what to configure and in the end failed. I'm not an advanced networker nor sysadmin, I must admit; however just to enable http3 it is too much. E.g. compare traefiks docker-compose http3 example (where one needs to just open the port) and what one needs to do with the helm chart

[the remark this answers is not in the document]
Comment (yoeluk) [reply traefik/Changelog.md:51]: @jnoordsij it looks like you are suggesting that there is a way to deploy this chart in Digital Ocean. I don't think that this is possible when you are exposing TCP and UDP ports. This chart selects all and any UDP port for a traefik-udp port and doesn't let us add a TCP port in that LB for healthcheck. I ended up cloning this chart locally and modifying the service template which is sad but while we are conversing here which seems will take a few weeks to sort out.

[the remark this answers is not in the document]
Comment (jnoordsij) [reply traefik/Changelog.md:51]: > @jnoordsij it looks like you are suggesting that there is a way to deploy this chart in Digital Ocean. I don't think that this is possible when you are exposing TCP and UDP ports. This chart selects all and any UDP port for a traefik-udp port and doesn't let us add a TCP port in that LB for healthcheck. I ended up cloning this chart locally and modifying the service template which is sad but while we are conversing here which seems will take a few weeks to sort out.

You're totally right, my apologies if my comments were unclear. The latest release of this chart does not support HTTP/3 with Digital Ocean setup 'out-of-the-box', that's why I started this MR in the first place. Using my local version of this chart with the proposed changes, I did manage to get it working now.

**Comment by mloiseleur** [comment]
@jnoordsij I reworked the part on http3 port in order to be simpler when enabling http3 and be more consistant with Traefik Proxy documentation. I made it in a single commit, so it's easy to see the diff and revert. What do you think about this approach ?

**Review by charlie-haley — APPROVED** [review]

**Review by mloiseleur — APPROVED** [review]

**Comment by jnoordsij** [comment]
> @jnoordsij I reworked the part on http3 port in order to be simpler when enabling http3 and be more consistant with Traefik Proxy documentation. I made it in a single commit, so it's easy to see the diff and revert.
> 
> For defaults, I'm in favor of staying with entrypoint port by default, so _443/UDP_, just like Traefik does, with proper documentation on how/why it can be an issue.
> 
> What do you think about this approach ?

Looks great to me, needing only a single flag is a lot better. Technically this removes the option to only enable HTTP/3 internally within your cluster, but that's probably not a big issue.

Using the same port number by default might be a little confusing due to the bug. For example, I think DigitalOcean will immediately issue an error when trying to apply this, but probably this will be handled gracefully. I'm slightly more afraid about upgrading existing installations, as improper merging might actually break existing TCP/443 ports. But as HTTP/3 is still marked as experimental, people should already be a little careful when trying to use it.

So all in all I think things look great now!

**Comment by smnbbrv** [comment]
When could we expect this to be released?
